enhancements/storage/csi-vsphere-operator.md

```diff
@@ -0,0 +1,159 @@
+* vSphere CSI driver provides new features such as - volume expansion, snapshotting and cloning which were previously unavailable with intree driver.
+
+### Goals
+
+* Create an operator to install the vSphere CSI driver.
+* Publish the driver and operator via default OCP build pipeline.
+* Enable creation of vSphere CSI storageclass, so as OCP users can start consuming vSphere CSI volumes without requiring further configuration.
+
+### Non-Goals
+
+* We don't intend to create a brand new driver with this KEP but merely want to deploy and configure upstream vSphere CSI driver - https://github.com/kubernetes-sigs/vsphere-csi-driver/
+
+
+## Proposal
+
+OCP ships with a vsphere-csi-driver-operator by default which is managed by [cluster-storage-operator](https://github.com/openshift/cluster-storage-operator/).
+vSphere CSI driver has few dependencies on installer though and they are:
+
+### Installer dependency
+
+* vSphere CSI driver requires HW version 15 on VMs that make up OCP cluster. Currently the rhcos OVA file Red Hat ships has default HW version configured to 13 and hence VM version should be updated.
+* Configuration of vSphere StorageClass requires knowledge of storage policy that was created in vCenter. Without this information - it is not possible to create a working storageClass for CSI driver.
+
+#### HW version of OCP VMs
+
+As mentioned above - vSphere CSI driver requires HW version 15 on all the VMs that make OCP cluster. Since, Openshift defaults to HW version 13 when a vSphere OCP cluster gets created - vSphere CSI driver isn't workable on the OCP cluster by default.
+
+To solve this following alternatives were considered:
+
+* We could provide manual instructions to the user about updating hardware version of their VMs. This won't be automatic but in some cases not even possible because some existing VMs can't be upgraded to HW version 15 in-place.
+* We could update installer to set HW version 15 - when appropriate hypervisor version(6.7u3 or higher) is detected.
+* For UPI install - we can additionally document required HW version while creating VMs.
+
+#### StoragePolicy configuration
+
+Currently while deploying Openshift a user can configure datastore used by OCP via install-config.yaml. vSphere CSI driver however can't use datastore directly and must be configured with vSphere storage policy. This can be done by an optional field in `vsphere.Platform` type of installer and if populated vsphere CSI driver operator will create storageClass with selected vSphere storage policy.
+
+One downside of having storage policy as an optional field is - when in-tree vSphere driver is removed from Kubernetes then without a default storage policy
+configured in OCP - such clusters will come up with no default storageClass installed and will require manual creation of storageClasses.
```

Comment:
Please note explicitly that we may need storage policy as mandatory in a future 4.x release. In the same way, datastore field may become optional and even deprecated + removed.

Reply:
datastore field we may still be needed to configure location where data files for VMs will be stored. cc @jcpowermac

Reply:
> datastore field we may still be needed to configure location where data files for VMs will be stored. cc @jcpowermac

Exactly it can't be removed. Its used in MAO and installer terraform.

Reply:
I don't like users configuring storage for PVs and for installer/machine API in two different ways. Would it be possible to create storagePolicy knowing the data store? Or the other way around? Or update terraform / machine API to use storagePolicy, if provided?

Reply:
theoretically it is possible to create a storagepolicy from datastore. We *could* create a tag and then apply the tag to datastore and then create a storagepolicy which selects datastores tagged with named tag.  If we are planning to do this via installer then I think support should be there in terraform to do that. 

@jsafrane were you thinking of doing this from operator?

Reply:
@gnufied we use tags already for virtualmachines, no reason why we can't add to include datastores.

https://github.com/openshift/installer/blob/master/data/data/vsphere/main.tf#L54-L70

Reply:
Tagging a datastore & creation of a storage policy on install time sounds great. Thinking about consequences:

* We will stick to datastore in install-config and users can't ever use their own storage policy for VMs.
* The default storage class will also use only that datastore.
  * If user wants a special storage policy, they must create their own StorageClass.

Works for me.

Reply:
* OCP  will also need permissions to create storage policies. This *may* be problematic.